File: CASE_Simple_Example_V4/Hardened/HAMR/src/main/component/HAMR/SW/FlightController_Impl_SW_FlightController_FlightController.scala
// #Sireum

package HAMR.SW

import org.sireum._
import HAMR._

// This file will not be overwritten so is safe to edit
object FlightController_Impl_SW_FlightController_FlightController {

  def initialise(api: FlightController_Impl_Initialization_Api): Unit = {
    // example api usage

    api.logInfo("Example info logging")
    api.logDebug("Example debug logging")
    api.logError("Example error logging")

  }

  def handle_FlightPlan(api: FlightController_Impl_Operational_Api, value : Base_Types.Bits): Unit = {
    api.logInfo("example handle_FlightPlan implementation")
    api.logInfo(s"received ${value}")
    // example api usage

    val apiUsage_FlightPlan: Option[Base_Types.Bits] = api.get_FlightPlan()
    api.logInfo(s"Received on FlightPlan: ${apiUsage_FlightPlan}")
  }

  def activate(api: FlightController_Impl_Operational_Api): Unit = { }

  def deactivate(api: FlightController_Impl_Operational_Api): Unit = { }

  def finalise(api: FlightController_Impl_Operational_Api): Unit = { }

  def recover(api: FlightController_Impl_Operational_Api): Unit = { }
}

File: CASE_Simple_Example_V4/Hardened-VM/HAMR_Simple_V4/src/c/ext-c/FlightController_Impl_SW_FlightController_FlightController/FlightController_Impl_SW_FlightController_FlightController.c
#include <FlightController_Impl_SW_FlightController_FlightController_api.h>
#include <FlightController_Impl_SW_FlightController_FlightController.h>
#include <ext.h>

// This file will not be overwritten so is safe to edit

Unit HAMR_Simple_V4_SW_FlightController_Impl_SW_FlightController_FlightController_initialise_(STACK_FRAME_ONLY) {
  DeclNewStackFrame(caller, "FlightController_Impl_SW_FlightController_FlightController.c", "", "HAMR_Simple_V4_SW_FlightController_Impl_SW_FlightController_FlightController_initialise_", 0);

  // examples of api setter and logging usage

  api_logInfo__HAMR_Simple_V4_SW_FlightController_Impl_SW_FlightController_FlightController(SF string("Example logInfo"));

  api_logDebug__HAMR_Simple_V4_SW_FlightController_Impl_SW_FlightController_FlightController(SF string("Example logDebug"));

  api_logError__HAMR_Simple_V4_SW_FlightController_Impl_SW_FlightController_FlightController(SF string("Example logError"));
}

Unit HAMR_Simple_V4_SW_FlightController_Impl_SW_FlightController_FlightController_finalise_(STACK_FRAME_ONLY) {
  DeclNewStackFrame(caller, "FlightController_Impl_SW_FlightController_FlightController.c", "", "HAMR_Simple_V4_SW_FlightController_Impl_SW_FlightController_FlightController_finalise_", 0);
}

Unit HAMR_Simple_V4_SW_FlightController_Impl_SW_FlightController_FlightController_timeTriggered_(STACK_FRAME_ONLY) {
  DeclNewStackFrame(caller, "FlightController_Impl_SW_FlightController_FlightController.c", "", "HAMR_Simple_V4_SW_FlightController_Impl_SW_FlightController_FlightController_timeTriggered_", 0);

  // examples of api getter usage

  uint8_t t0[numBytes_HAMR_Simple_V4_Data_Types_Mission];
  size_t t0_numBits;
  if(api_get_FlightPlan__HAMR_Simple_V4_SW_FlightController_Impl_SW_FlightController_FlightController(SF &t0_numBits, t0)) {
    // sanity check
    sfAssert(SF (Z) t0_numBits == numBits_HAMR_Simple_V4_Data_Types_Mission, "numBits received does not match expected");

    DeclNewString(FlightPlan_str);
    String__append(SF (String) &FlightPlan_str, string("Received on FlightPlan: "));
    byte_array_string(SF (String) &FlightPlan_str, t0, numBytes_HAMR_Simple_V4_Data_Types_Mission);
    api_logInfo__HAMR_Simple_V4_SW_FlightController_Impl_SW_FlightController_FlightController(SF (String) &FlightPlan_str);
  }

  if(api_get_Alert__HAMR_Simple_V4_SW_FlightController_Impl_SW_FlightController_FlightController(SF_LAST )){
    String Alert_str = string("Received event on Alert");
    api_logInfo__HAMR_Simple_V4_SW_FlightController_Impl_SW_FlightController_FlightController(SF Alert_str);
  }
}
